### DISCUSSION TITLE
Access rights for invited users

### DISCUSSION THREAD

**Discussion by amos-kibet** [3Stones-io/omedis#427] [Q&A] [answered]
When an invited user creates an account, we don't create any access rights for them. At the least, I think they should have read access to the organisation resource because we have these plugs that require the user to have read access:
```elixir
# plug OmedisWeb.Plugs.OrganisationsCount
def call(conn, _opts) do
    organisations_count =
      case conn.assigns[:current_user] do
        %User{} = user ->
          Ash.count!(Organisation, actor: user)

        _ ->
          0
      end

    assign(conn, :organisations_count, organisations_count)
  end
```

```elixir
# plug OmedisWeb.Plugs.CurrentOrganisation
def call(conn, _opts) do
    current_organisation =
      with %User{current_organisation_id: current_organisation_id}
           when not is_nil(current_organisation_id) <-
             conn.assigns[:current_user],
           {:ok, organisation} <-
             Organisation.by_id(current_organisation_id, actor: conn.assigns.current_user) do
        organisation
      else
        _ ->
          nil
      end

    assign(conn, :current_organisation, current_organisation)
  end
```

Maybe we don't even need the `OrganisationsCount` plug anymore, since the user will always belong to one organisation.

Since we have a **Users** group for every organisation, that has read access to select resources (currently all resources), I think the invited user should be added to that group. 

Lastly, what do we do with users who register directly, and not via invitation? With the new registration flow that @kagure-nyakio is working on, I guess she can add the new user to the **Users** group that gets created when the organisation for that user is created.

Let me know what you think, @almirsarajcic , @wintermeyer , @kagure-nyakio

**Comment by DenisMGojak** [comment, marked as answer]
Cursor has an opinion: 
I'll help analyze this situation and suggest some improvements. Let's break this down into parts:
Invited Users
Direct Registration
Plug Optimization
Here are my suggested changes:
For invited users, we should add them to the Users group when accepting the invitation:
```elixir
# In the accept action
def accept(invitation) do
  invitation
  |> add_argument(:password, :string, allow_nil?: false)
  |> add_argument(:password_confirmation, :string, allow_nil?: false)
  |> validate_password_confirmation()
  |> validate_acceptance()
  |> build_user()
  |> add_to_users_group()  # New step
end

# Add this private function
defp add_to_users_group(changeset) do
  after_transaction(changeset, fn _invitation, %{user: user} = result ->
    # Get the Users group for the organisation
    users_group = Groups.get_users_group!(user.organisation_id)
    
    # Add user to the group
    {:ok, _} = Groups.add_user_to_group(users_group.id, user.id)
    
    {:ok, result}
  end)
end
```
For direct registration, we should modify the organisation creation to include creating and linking the Users group:
```elixir
# In the create action or wherever organisation creation happens
def create_organisation_with_user(changeset) do
  after_transaction(changeset, fn organisation, result ->
    # Create Users group
    {:ok, users_group} = Groups.create_users_group(organisation.id)
    
    # Add the creating user to the Users group
    {:ok, _} = Groups.add_user_to_group(users_group.id, result.user.id)
    
    {:ok, result}
  end)
end
```
Regarding the plugs, I agree that OrganisationsCount might not be necessary anymore if users will always belong to exactly one organisation. Here's how we could simplify the CurrentOrganisation plug:
```elixir
def call(conn, _opts) do
  current_organisation =
    case conn.assigns[:current_user] do
      %User{current_organisation_id: org_id} when not is_nil(org_id) ->
        # User will always have read access through Users group
        {:ok, org} = Organisation.by_id(org_id, actor: conn.assigns.current_user)
        org
      _ ->
        nil

end
```

A few additional recommendations:
You can remove the OrganisationsCount plug from your pipeline since it's no longer needed.
Consider adding a test to ensure users always get added to the Users group:

```elixir
test "accepting invitation adds user to Users group" do
  invitation = create_invitation()
  {:ok, result} = Invitation.accept(invitation, %{password: "valid_password", password_confirmation: "valid_password"})
  
  users_group = Groups.get_users_group!(result.user.organisation_id)
  assert Groups.user_in_group?(users_group.id, result.user.id)
end
````
This approach ensures that:
Invited users get proper access through the Users group
Directly registered users get proper access through the Users group created with their organisation
The code is simpler by removing unnecessary plugs
Access control is consistent across all user types
Let me know if you'd like me to elaborate on any part of this solution!

**Reply by amos-kibet** [reply]
Thanks @DenisMGojak.

**Comment by almirsarajcic** [comment]
@amos-kibet, invitee needs to be added to the `Users` group.
I've created a task for this: https://github.com/3Stones-io/omedis/issues/428.

For your other question, when the user registers without an invitation, they are creating a new organisation which we already handle so there's nothing to add there.

**Comment by amos-kibet** [comment]
When creating an invitation on the UI, the user has an option of adding the invitee to the groups of their choice, including the **Users** group. ([I fixed the seeds](https://github.com/3Stones-io/omedis/pull/414#issuecomment-2514810918) 💪 ) 
<img width="1502" alt="Screenshot 2024-12-09 at 15 24 29" src="https://github.com/user-attachments/assets/d3e64ef2-f536-4440-8f0d-b0e6a0afd075">

But I don't think that has been setup properly. Adding a user to a group involves creating a group_membership, and I don't see that in the invited user registration flow.

Also, I think adding invitees to the **Users** group should be mandatory, otherwise they won't have access to the organisation's resources.
@almirsarajcic, I have implemented an `after_action` hook on user registration, so invited users will be added to the **Users** group, but I think we should fix the invitation form, the part of selecting groups. What do you think?

**Reply by almirsarajcic** [reply]
Good catch, you're right. Can you create a new task for that?
Let's try this approach: `Users` should be checked and disabled so that we can't uncheck the group (it should be appended on the backend when submitting the form because disabled fields aren't included in params).

**Reply by amos-kibet** [reply]
I have created the issue (#432) in the **Ready** column.

**Reply by almirsarajcic** [reply]
Thanks
